api/src/main/java/io/opencensus/trace/propagation/TextFormat.java

```diff
@@ -0,0 +1,184 @@
+   * @param carrier holds propagation fields. For example, an outgoing message or http request.
+   * @param setter invoked for each propagation key to add or remove.
+   */
+  public abstract <C> void inject(SpanContext spanContext, C carrier, Setter<C> setter);
+
+  /**
+   * Class that allows to be saved as a constant to avoid runtime allocations.
+   *
+   * @param <C> carrier of propagation fields, such as an http request
+   */
+  public abstract static class Setter<C> {
+
+    /**
+     * Replaces a propagated field with the given value.
+     *
+     * <p>For example, a setter for an {@link java.net.HttpURLConnection} would be the method
+     * reference {@link java.net.HttpURLConnection#addRequestProperty(String, String)}
+     *
+     * @param carrier holds propagation fields. For example, an outgoing message or http request.
+     * @param key the key of the field.
+     * @param value the value of the field.
+     */
+    public abstract void put(C carrier, String key, String value);
+  }
+
+  /**
+   * Extracts the span context from upstream. For example, as http headers.
+   *
+   * <p>If no particular format is known to be always received on from the wire it is recommended to
+   * use {@code Encoding.ALL}.
+   *
+   * @param carrier holds propagation fields. For example, an outgoing message or http request.
+   * @param getter invoked for each propagation key to get.
+   * @throws SpanContextParseException if the input is invalid
+   */
+  public abstract <C> SpanContext extract(C carrier, Getter<C> getter)
+      throws SpanContextParseException;
+
+  /**
+   * Class that allows to be saved as a constant to avoid runtime allocations.
+   *
+   * @param <C> carrier of propagation fields, such as an http request
+   */
+  public abstract static class Getter<C> {
+
+    /**
+     * Returns the first value of the given propagation {@code key} or returns {@code null}.
+     *
+     * @param carrier carrier of propagation fields, such as an http request
+     * @param key the key of the field.
+     * @return the first value of the given propagation {@code key} or returns {@code null}.
+     */
+    @Nullable
+    public abstract String get(C carrier, String key);
+  }
+
+  /**
+   * Returns the no-op implementation of the {@code TextFormat}.
+   *
+   * @return the no-op implementation of the {@code TextFormat}.
+   */
+  static TextFormat getNoopTextFormat() {
+    return NOOP_TEXT_FORMAT;
+  }
+
+  private static final class NoopTextFormat extends TextFormat {
+
+    private NoopTextFormat() {}
+
+    @Override
+    public List<String> fields() {
+      return Collections.emptyList();
+    }
+
+    @Override
+    public <C> void inject(SpanContext spanContext, C carrier, Setter<C> setter) {
+      checkNotNull(spanContext, "spanContext");
+      checkNotNull(setter, "setter");
+    }
+
+    @Override
+    public <C> SpanContext extract(C carrier, Getter<C> getter) {
+      checkNotNull(getter, "getter");
```

Comment:
Should this method also check the carrier for null?  If it can be null, then I think it would help to mention what can be null in the Javadocs above.

Reply:
yeah carrier shouldn't be allowed to be null

Reply:
Done.